startup/92-optimization.py

```diff
@@ -1,70 +1,124 @@
-#
-# In [9]:
-
-# ^^^ solved by rewriting the watch function (no globals)
-
-# TODO:
-# 1) have a live plot for the fitness (convergence plot)
-# 2) improve the algorithm to optimize the travel path (should not go back to the same point)
-# 3) Usability:
-#    - convenient I/O to the saved data (pandas dataframes, sorting, ...)
-#    - clean up prints and better output log
-#    - add metadata about the samples
-#    - plot trajectory
+import sys
+import textwrap
+import traceback
+
+from tabulate import tabulate
+
+
+def _extract_tb():
+    shared_kwargs = {"tablefmt": "rounded_grid"}
+    traceback_kwargs = {**shared_kwargs}
+    maxcolwidths = {}
+    two_borders = 2 * 2
+    indent = 4
+
+    exc_type, exc_value, exc_traceback = sys.exc_info()
+    tb = traceback.format_tb(exc_traceback)
+
+    max_len = max([len(line) for line in tb])
+    term_width = os.get_terminal_size().columns
+
+    tb = "".join(tb)
+
+    if max_len >= term_width - two_borders:
+        traceback_kwargs.update({"maxcolwidths": term_width - two_borders})
+
+    print(tabulate([[tb]], headers=("Traceback",), **traceback_kwargs))
+    print(
+        textwrap.indent(
+            tabulate(
+                [[exc_value]],
+                headers=("Exception",),
+                maxcolwidths=(term_width - two_borders - indent),
+                **shared_kwargs,
+            ),
+            prefix=" " * indent,
+        )
+    )
+
+
+if auto_alignment_mode():
+    try:
+        # Imports for bloptools < v0.0.2.
+        from bloptools.de_opt_utils import run_hardware_fly
+        from bloptools.de_optimization import optimization_plan
+    except (ImportError, ModuleNotFoundError):
+        _extract_tb()
+
+        print("\nFalling back to a newer version of bloptools...\n")
+        # The modules were moved to the `de` package in https://github.com/NSLS-II/bloptools/pull/5.
+        # To be released as v0.1.0 or newer.
+        try:
+            from bloptools import gp
+            from bloptools.de.de_opt_utils import run_hardware_fly
+            from bloptools.de.de_optimization import optimization_plan
+        except (ImportError, ModuleNotFoundError):
+            _extract_tb()
+            print(f"\nContinuing without bloptools...\n")
+
+    ###########################################################################
+    #                            DE optimization                              #
+    ###########################################################################
+
+    motor_dict = {
+        sample_stage.x.name: {"position": sample_stage.x},
+        sample_stage.y.name: {"position": sample_stage.y},
+        # sample_stage.z.name: {'position': sample_stage.z},
+    }
+
+    bound_vals = [(43, 44), (34, 35)]
+    motor_bounds = {}
+    motor_dict_keys = list(motor_dict.keys())
+    for k in range(len(motor_dict_keys)):
+        motor_bounds[motor_dict_keys[k]] = {"position": [bound_vals[k][0], bound_vals[k][1]]}
```

Comment:
this does not seem to be used

Reply:
These parameters are supposed to work with the DE optimization (not currently used). Should we comment these lines out or should we remove them?

Reply:
I commented them out in 0704c6e4f8d177bf7ac97667dc48e46b94d182da.